File: skills/python-scaffold/src/python_scaffold/command_output.py
import os
import shlex
import subprocess
from collections.abc import Callable, Mapping
from pathlib import Path
from typing import BinaryIO
from uuid import uuid4

from platformdirs import user_log_path

FAILURE_TAIL_LINES = 20
FAILURE_TAIL_BYTES = 64 * 1024


class CommandSession:
    def __init__(
        self,
        *,
        log_directory: Path | None = None,
        identifier: str | None = None,
        write_line: Callable[[str], None] = print,
    ) -> None:
        directory = log_directory or command_log_directory()
        self.directory = directory / (identifier or str(uuid4()))
        self._write_line = write_line
        self._command_count = 0
        self._reported = False

    def run(self, command: tuple[str, ...], cwd: Path) -> None:
        self._command_count += 1
        prefix = f"{self._command_count:03d}"
        display = shlex.join(command)
        stdout_path = self.directory / f"{prefix}.stdout.log"
        stderr_path = self.directory / f"{prefix}.stderr.log"
        index_path = self.directory / "commands.log"
        self.directory.mkdir(mode=0o700, parents=True, exist_ok=True)
        self._append_private(
            index_path,
            "\n".join(
                [
                    f"[{prefix}]",
                    f"$ {display}",
                    f"cwd: {cwd}",
                    f"stdout: {stdout_path}",
                    f"stderr: {stderr_path}",
                    "",
                ]
            ),
        )

        try:
            with (
                self._open_private(stdout_path) as stdout,
                self._open_private(stderr_path) as stderr,
            ):
                process = subprocess.Popen(command, cwd=cwd, stdout=stdout, stderr=stderr)
                exit_code = process.wait()
        except OSError as error:
            self._append_private(index_path, "exit: unavailable\n\n")
            self._reported = True
            raise RuntimeError(
                self._failure_message(display, None, stdout_path, stderr_path, str(error))
            ) from error

        self._append_private(index_path, f"exit: {exit_code}\n\n")
        if exit_code != 0:
            self._reported = True
            raise RuntimeError(self._failure_message(display, exit_code, stdout_path, stderr_path))
        self._write_line(f"PASS {display}")

    def report(self) -> None:
        if self._command_count > 0 and not self._reported:
            self._reported = True
            self._write_line(f"Full command logs: {self.directory}")

    def _failure_message(
        self,
        command: str,
        exit_code: int | None,
        stdout_path: Path,
        stderr_path: Path,
        fallback: str = "",
    ) -> str:
        stdout_tail = self._read_tail(stdout_path)
        stderr_tail = self._read_tail(stderr_path)
        details = [
            section
            for section in (
                f"[stdout tail]\n{stdout_tail}" if stdout_tail else "",
                f"[stderr tail]\n{stderr_tail}" if stderr_tail else "",
            )
            if section
        ]
        if not details:
            details.append(fallback or "No command output captured.")
        status = exit_code if exit_code is not None else "unavailable"
        return "\n".join(
            [
                f"Command failed: {command} (exit {status})",
                *details,
                f"Full command logs: {self.directory}",
            ]
        )

    @staticmethod
    def _open_private(path: Path) -> BinaryIO:
        descriptor = os.open(path, os.O_WRONLY | os.O_CREAT | os.O_EXCL, 0o600)
        return os.fdopen(descriptor, "wb")

    @staticmethod
    def _append_private(path: Path, content: str) -> None:
        descriptor = os.open(path, os.O_WRONLY | os.O_CREAT | os.O_APPEND, 0o600)
        with os.fdopen(descriptor, "a", encoding="utf-8") as log:
            log.write(content)

    @staticmethod
    def _read_tail(path: Path) -> str:
        if not path.exists():
            return ""
        with path.open("rb") as log:
            log.seek(0, os.SEEK_END)
            length = min(log.tell(), FAILURE_TAIL_BYTES)
            if length == 0:
                return ""
            log.seek(-length, os.SEEK_END)
            output = log.read().decode("utf-8", errors="replace")
        return "\n".join(output.splitlines()[-FAILURE_TAIL_LINES:])


def command_log_directory(environment: Mapping[str, str] = os.environ) -> Path:
    configured = environment.get("AGENT_KIT_LOG_DIR")
    root = Path(configured).expanduser() if configured else user_log_path("agent-kit")
    return root / "scaffolds/python"

File: skills/typescript-scaffold/src/command-output.ts
import { appendFile, mkdir, open } from "node:fs/promises";
import { spawn } from "node:child_process";
import { randomUUID } from "node:crypto";
import { join } from "node:path";
import envPaths from "env-paths";

const FAILURE_TAIL_LINES = 20;
const FAILURE_TAIL_BYTES = 64 * 1024;

export type CommandRunner = (
  command: string,
  args: string[],
  options: { cwd: string },
) => Promise<void>;

interface CommandSessionOptions {
  logDirectory?: string;
  identifier?: string;
  writeLine?: (line: string) => void;
}

interface LoggedCommand {
  display: string;
  stdoutPath: string;
  stderrPath: string;
}

class LoggedCommandError extends Error {}

export interface CommandSession {
  run: CommandRunner;
  readStdout(command: string, args: string[], options: { cwd: string }): Promise<string>;
  report(): void;
}

export function createCommandSession(
  options: CommandSessionOptions = {},
): CommandSession {
  const logDirectory = options.logDirectory ?? resolveCommandLogDirectory();
  const sessionDirectory = join(logDirectory, options.identifier ?? randomUUID());
  const writeLine = options.writeLine ?? ((line: string) => console.log(line));
  let commandCount = 0;
  let reported = false;

  const execute = async (
    command: string,
    args: string[],
    commandOptions: { cwd: string },
  ): Promise<LoggedCommand> => {
    commandCount += 1;
    try {
      return await executeLoggedCommand(
        command,
        args,
        commandOptions,
        sessionDirectory,
        commandCount,
        writeLine,
      );
    } catch (error) {
      if (error instanceof LoggedCommandError) reported = true;
      throw error;
    }
  };

  return {
    async run(command, args, commandOptions) {
      await execute(command, args, commandOptions);
    },
    async readStdout(command, args, commandOptions) {
      const logged = await execute(command, args, commandOptions);
      return readTail(logged.stdoutPath);
    },
    report() {
      if (commandCount > 0 && !reported) {
        reported = true;
        writeLine(`Full command logs: ${sessionDirectory}`);
      }
    },
  };
}

export function resolveCommandLogDirectory(
  environment: NodeJS.ProcessEnv = process.env,
): string {
  const root = environment.AGENT_KIT_LOG_DIR
    ?? envPaths("agent-kit", { suffix: "" }).log;
  return join(root, "scaffolds", "typescript");
}

async function executeLoggedCommand(
  command: string,
  args: string[],
  options: { cwd: string },
  sessionDirectory: string,
  commandNumber: number,
  writeLine: (line: string) => void,
): Promise<LoggedCommand> {
  await mkdir(sessionDirectory, { recursive: true, mode: 0o700 });
  const prefix = String(commandNumber).padStart(3, "0");
  const stdoutPath = join(sessionDirectory, `${prefix}.stdout.log`);
  const stderrPath = join(sessionDirectory, `${prefix}.stderr.log`);
  const display = formatCommand(command, args);
  const indexPath = join(sessionDirectory, "commands.log");
  await appendPrivateFile(indexPath, [
    `[${prefix}]`,
    `$ ${display}`,
    `cwd: ${options.cwd}`,
    `stdout: ${stdoutPath}`,
    `stderr: ${stderrPath}`,
  ].join("\n") + "\n");

  const stdout = await open(stdoutPath, "wx", 0o600);
  let stderr;
  try {
    stderr = await open(stderrPath, "wx", 0o600);
  } catch (error) {
    await stdout.close();
    throw error;
  }

  let commandError: unknown;
  try {
    await spawnCommand(command, args, options.cwd, stdout.fd, stderr.fd);
  } catch (error) {
    commandError = error;
  } finally {
    await Promise.all([stdout.close(), stderr.close()]);
  }

  if (commandError) {
    const error = commandError;
    const exitCode = errorNumber(error, "exitCode");
    await appendPrivateFile(indexPath, `exit: ${exitCode ?? "unavailable"}\n\n`);
    const [stdoutTail, stderrTail] = await Promise.all([
      readTail(stdoutPath),
      readTail(stderrPath),
    ]);
    throw new LoggedCommandError(
      failureMessage(display, exitCode, stdoutTail, stderrTail, error, sessionDirectory),
      { cause: error },
    );
  }

  await appendPrivateFile(indexPath, "exit: 0\n\n");
  writeLine(`PASS ${display}`);
  return { display, stdoutPath, stderrPath };
}

async function appendPrivateFile(path: string, content: string): Promise<void> {
  await appendFile(path, content, { encoding: "utf8", mode: 0o600 });
}

async function spawnCommand(
  command: string,
  args: string[],
  cwd: string,
  stdout: number,
  stderr: number,
): Promise<void> {
  const child = spawn(command, args, {
    cwd,
    stdio: ["ignore", stdout, stderr],
    windowsHide: true,
  });
  await new Promise<void>((resolve, reject) => {
    child.once("error", reject);
    child.once("close", (exitCode, signal) => {
      if (exitCode === 0) {
        resolve();
        return;
      }
      reject(Object.assign(
        new Error(signal ? `Command terminated by ${signal}` : `Command exited with ${exitCode}`),
        { exitCode: exitCode ?? undefined },
      ));
    });
  });
}

async function readTail(path: string): Promise<string> {
  const file = await open(path, "r");
  try {
    const metadata = await file.stat();
    const length = Math.min(metadata.size, FAILURE_TAIL_BYTES);
    if (length === 0) return "";
    const buffer = Buffer.alloc(length);
    await file.read(buffer, 0, length, metadata.size - length);
    return buffer.toString("utf8").trimEnd()
      .split(/\r?\n/).slice(-FAILURE_TAIL_LINES).join("\n");
  } finally {
    await file.close();
  }
}

function failureMessage(
  command: string,
  exitCode: number | undefined,
  stdoutTail: string,
  stderrTail: string,
  error: unknown,
  sessionDirectory: string,
): string {
  const details = [
    stdoutTail ? `[stdout tail]\n${stdoutTail}` : "",
    stderrTail ? `[stderr tail]\n${stderrTail}` : "",
  ].filter(Boolean);
  if (details.length === 0) details.push(shortError(error));
  return [
    `Command failed: ${command} (exit ${exitCode ?? "unavailable"})`,
    ...details,
    `Full command logs: ${sessionDirectory}`,
  ].join("\n");
}

function shortError(error: unknown): string {
  if (error && typeof error === "object") {
    const shortMessage = Reflect.get(error, "shortMessage");
    if (typeof shortMessage === "string") return shortMessage;
  }
  return error instanceof Error ? error.message : String(error);
}

function errorNumber(error: unknown, key: string): number | undefined {
  if (!error || typeof error !== "object") return undefined;
  const value = Reflect.get(error, key);
  return typeof value === "number" ? value : undefined;
}

function formatCommand(command: string, args: string[]): string {
  return [command, ...args].map(formatArgument).join(" ");
}

function formatArgument(value: string): string {
  return /^[a-zA-Z0-9_./:@%+=,-]+$/.test(value) ? value : JSON.stringify(value);
}
